### PR TITLE
feat: cassandra-googleads integration

### PR DESCRIPTION
**Pull request by shubhank-v**
### What changes are proposed in this PR?
---
* This PR integrated Cassandra to capture the state of google ads gclid.

### How was this tested?
---
* Unit test

### CODE REVIEW

File: hip_data_tools/etl/athena_to_googleads.py [hunk 1]
```diff
@@ -172,15 +174,17 @@ def _chunk_batches(self, lst: List[dict]) -> List[List[dict]]:
...
+        self._state_manager_connect()
+        ready_data, verification_issues = self._verify_data_before_upsert(data_dict)
+        data_dict_batches = self._chunk_batches(ready_data)
+        # data_dict_batches = self._chunk_batches(data_dict)
         successes = []
         failures = []
         verification_issues = []
 
         for data_batch in data_dict_batches:
+            data_to_process, processing_issue = self._mark_processing(data_batch)
```
[the remark this answers is not in the document]
Comment (shubhank-v) [reply hip_data_tools/etl/athena_to_googleads.py:186]: fixed

File: hip_data_tools/etl/athena_to_googleads.py [hunk 2]
```diff
@@ -41,7 +43,7 @@ class AthenaToGoogleAdsOfflineConversionSettings(AthenaToDataFrameSettings):
 
 
 def _get_record_signature(record: dict):
-    return f"{record['googleClickId']}||{record['conversionName']}||{record['conversionTime']}"
+    return f"{record['gclid']}||conversion_attempt||{record['conversion_date_time']}"
```
[the remark this answers is not in the document]
Comment (shubhank-v) [reply hip_data_tools/etl/athena_to_googleads.py:46]: Yes, Conversion_attempt is hardcoded in the source.

File: hip_data_tools/etl/athena_to_googleads.py [hunk 3]
```diff
@@ -198,6 +202,68 @@ def _process_data_frame(self, data_frame) -> Tuple[list, list, list]:
...
+
+    def _verify_data_before_upsert(self, data: List[dict]) -> (List[dict], List[dict]):
+        data, issues = map(list, zip(*[self._sanitise_data(dat) for dat in data]))
+
+        if len(issues) > 0:
+            LOG.warning("Issues found in verification, number of issues: %i",
+                        len(issues))
+
+        # Remove None from the List
+        return [i for i in data if i], [i for i in issues if i]
```
[the remark this answers is not in the document]
Comment (shubhank-v) [reply hip_data_tools/etl/athena_to_googleads.py:241]: I want to test this in feature branch, after better understanding of data I'll hand the None in logging

File: hip_data_tools/etl/athena_to_googleads.py [hunk 4]
```diff
@@ -172,17 +174,19 @@ def _chunk_batches(self, lst: List[dict]) -> List[List[dict]]:
...
     def _process_data_frame(self, data_frame) -> Tuple[list, list, list]:
         data_dict = self._data_frame_to_destination_dict(data_frame)
-        # self._state_manager_connect()
-        # ready_data, verification_issues = self._verify_data_before_upsert(data_dict)
-        # data_dict_batches = self._chunk_batches(ready_data)
-        data_dict_batches = self._chunk_batches(data_dict)
+        self._state_manager_connect()
+        ready_data, verification_issues = self._verify_data_before_upsert(data_dict)
+        data_dict_batches = self._chunk_batches(ready_data)
+        # data_dict_batches = self._chunk_batches(data_dict)
```
[the remark this answers is not in the document]
Comment (shubhank-v) [reply hip_data_tools/etl/athena_to_googleads.py:180]: removed unused code

**Review by samithdisal — APPROVED** [review]

**Pull request merged**